### PR TITLE
Store the coerced base value in the BaseMutation

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
Port of https://github.com/firebase/firebase-android-sdk/pull/564

The PR changes what we store in a BaseMutation. With the changes here, every transform operation computes what it needs to store in the BaseMutation. For Numeric increment, this is either an existing numeric value or the IntegerValue.of(0), replacing all invalid base values at time of creation.

### CODE REVIEW

File: packages/firestore/src/model/field_value.ts [hunk 1]
```diff
@@ -605,6 +606,31 @@ export class ObjectValue extends FieldValue {
     return field;
   }
 
+  /** Recursively extracts the FieldPaths that are set in this ObjectValue. */
```
Comment (Hui-Wu) [inline packages/firestore/src/model/field_value.ts]: I feel like the comments here can be elaborated, something like:

Returns a `FieldMask` built from all `FieldPath`s starting from this `ObjectValue`, including paths from nested objects.

File: packages/firestore/src/model/mutation.ts [hunk 2]
```diff
@@ -302,17 +276,25 @@ export abstract class Mutation {
...
 
-  abstract isEqual(other: Mutation): boolean;
-
   /**
-   * If applicable, returns the field mask for this mutation. Fields that are
-   * not included in this field mask are not modified when this mutation is
-   * applied. Mutations that replace the entire document return 'null'.
+   * If applicable, returns the base value to persist with this mutation. If a
+   * base value is provided, the mutation is always applied to this base value,
+   * even if document has already beenupdated.
```
Comment (Hui-Wu) [inline packages/firestore/src/model/mutation.ts]: beenupdated -> been updated

File: packages/firestore/src/model/mutation.ts [hunk 3]
```diff
@@ -302,17 +276,25 @@ export abstract class Mutation {
...
+   * behavior for non-idempotent transforms and allow us to return the same
+   * latency-compensated value even if the backend has already applied the
+   * mutation. The base value is null for idempotent mutations, as they can be
+   * re-played even if the backend has already applied them.
+   *
+   * @return a base value to store along with the mutation, or null for
+   * idempotent mutations.
    */
-  abstract get fieldMask(): FieldMask | null;
+  abstract extractBaseValue(maybeDoc: MaybeDocument | null): ObjectValue | null;
```
Comment (Hui-Wu) [inline packages/firestore/src/model/mutation.ts]: As a reader who don't have enough knowledge, `BaseValue` is a little vague. I could not understand it until after reading the comments. I feel like maybe we should try to come up with a better name, my attempt is `computeBaseValueToApplyMutation`... a bit wordy.

File: packages/firestore/src/model/mutation.ts [hunk 4]
```diff
@@ -302,17 +276,25 @@ export abstract class Mutation {
     localWriteTime: Timestamp
   ): MaybeDocument | null;
 
-  abstract isEqual(other: Mutation): boolean;
-
   /**
-   * If applicable, returns the field mask for this mutation. Fields that are
-   * not included in this field mask are not modified when this mutation is
-   * applied. Mutations that replace the entire document return 'null'.
+   * If applicable, returns the base value to persist with this mutation. If a
```
Comment (Hui-Wu) [inline packages/firestore/src/model/mutation.ts]: When is this not `applicable`?

File: packages/firestore/src/model/mutation.ts [hunk 5]
```diff
@@ -600,22 +579,29 @@ export class TransformMutation extends Mutation {
...
+  extractBaseValue(maybeDoc: MaybeDocument | null): ObjectValue | null {
+    let baseObject: ObjectValue | null = null;
     for (const fieldTransform of this.fieldTransforms) {
-      if (!fieldTransform.isIdempotent) {
-        return false;
+      const existingValue =
+        maybeDoc instanceof Document
+          ? maybeDoc.field(fieldTransform.field)
+          : undefined;
+      const coercedValue = fieldTransform.transform.computeBaseValue(
```
Comment (Hui-Wu) [inline packages/firestore/src/model/mutation.ts]: Not a comment but a question: where is the implementation of `computeBaseValue` defined. I cannot find it..

File: packages/firestore/src/model/field_value.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: Done (went with your version).

File: packages/firestore/src/model/mutation.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Replaced with "If this mutation is not idempotent"

File: packages/firestore/src/model/mutation.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Fixed

File: packages/firestore/src/model/mutation.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: I can't come up with a better term that is still concise. For better or for worse, the "base" concept is already used in our serialization (https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/local/indexeddb_schema.ts#L429) and may just be one of the core concepts that someone has to ramp up on before they can understand how our mutations are applied. I agree that it would be nicer if the name alone conveyed what this value is used for, but I can't come up with anything short enough to win me over.

File: packages/firestore/src/model/mutation.ts (hunk 5 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: It's added in this PR: https://github.com/firebase/firebase-js-sdk/pull/1928/files#diff-eb38c9da3263f013aac932a74339324dR228

Comment (Hui-Wu) [reply packages/firestore/src/model/mutation.ts]: Ah, I overlooked..thx.

**Comment by schmidt-sebastian** [comment]
Adding @wilhuff for Owners approval.

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 6]
```diff
@@ -209,6 +209,14 @@ apiDescribe('Server Timestamps', (persistence: boolean) => {
   });
 
   it('can return previous value', () => {
+    // The following update includes an update of the nested map "deep", which
+    // updates it to contain a single ServerTimestamp. As such, the update is
+    // split into two mutations: One that sets "deep" to an empty map and
+    // overwrites the previous ServerTimestamp value and a second ransform that
```
Comment (wilhuff) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: s/ransform/transform/

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 7]
```diff
@@ -209,6 +209,14 @@ apiDescribe('Server Timestamps', (persistence: boolean) => {
   });
 
   it('can return previous value', () => {
+    // The following update includes an update of the nested map "deep", which
```
Comment (wilhuff) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: "The following update includes an update" sounds wrong. Maybe you meant "The following test includes an update"?

Also "As such" doesn't work here. I think what you're trying to say is that the implementation of the update call is to split it into two mutations ...

Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: Removed "as such" and replaced with "This update".
